src/backend/distributed/shardsplit/shardsplit_decoder.c

```diff
@@ -62,6 +77,80 @@ _PG_output_plugin_init(OutputPluginCallbacks *cb)
 	/* actual pgoutput callback will be called with the appropriate destination shard */
 	pgoutputChangeCB = cb->change_cb;
 	cb->change_cb = split_change_cb;
+	cb->filter_by_origin_cb = replication_origin_filter_cb;
+}
+
+
+/*
+ * replication_origin_filter_cb call back function filters out publication of changes
+ * originated from any other node other than the current node. This is
+ * identified by the "origin_id" of the changes. The origin_id is set to
+ * a non-zero value in the origin node as part of WAL replication.
+ */
+static bool
+replication_origin_filter_cb(LogicalDecodingContext *ctx, RepOriginId origin_id)
+{
+	if (origin_id != InvalidRepOriginId)
+	{
+		return true;
+	}
+	return false;
+}
+
+
+/*
+ * PublishChangesIfCdcSlot checks if the current slot is a CDC slot. If so, it publishes
+ * the changes as the change for the distributed table instead of shard.
+ * If not, it returns false. It also skips the Citus metadata tables.
+ */
+static bool
+PublishChangesIfCdcSlot(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
+						Relation relation, ReorderBufferChange *change)
+{
+	char *replicationSlotName = ctx->slot->data.name.data;
+
+	/* Check if the replication slot is CITUS_CDC_SLOT*/
+	if (replicationSlotName != NULL &&
+		strncmp(replicationSlotName, CITUS_SHARD_PREFIX_SLOT, strlen(
+					CITUS_SHARD_PREFIX_SLOT)) != 0)
+	{
+		/* Skip publishing changes for system relations in pg_catalog*/
+		if (relation->rd_rel->relnamespace == PG_CATALOG_NAMESPACE)
+		{
+			return true;
+		}
+
+		char *shardRelationName = RelationGetRelationName(relation);
+		uint64 shardId = ExtractShardIdFromTableName(shardRelationName, true);
+		if (shardId != INVALID_SHARD_ID && ShardExists(shardId))
+		{
+			if (ReferenceTableShardId(shardId))
+			{
+				/*For reference tables, publish the changes only from the coordinator node. */
+				if (!IsCoordinator())
+				{
+					int nodeID = GetLocalNodeId();
+					elog(LOG,
```

Comment:
LOG seems too aggressive, since this goes to server log by default for every write

also, can we use ereport and move the GetLocalNodeId call inline? (such that it does not get called if the log level is higher)

Reply:
Removed this LOG, this is just for debugging.